core/vault/vault.go

```diff
@@ -0,0 +1,254 @@
+		Vault: base64.RawStdEncoding.EncodeToString(encVf),
+		Vsk:   base64.RawStdEncoding.EncodeToString(vsk),
+	}
+	reqJSON, err := json.Marshal(storeRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := http.Post(
+		v.vaultAPIURL+"/vaults",
+		"application/json",
+		bytes.NewBuffer(reqJSON),
+	)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	_, err = parseAPIResponse(resp)
+	if err != nil {
+		return nil, err
+	}
+
+	return storeRequest, nil
+}
+
+func (v *vault) Retrieve(uuid string, passphrase string) ([]VaultItem, error) {
+	// Compute vault key
+	vk := v.computeVk(uuid, passphrase, VkVersion1)
+
+	// Compute vault service key
+	vsk := v.computeVsk(vk, passphrase, VkVersion1)
+
+	// Send retrieve request to vault service
+	reqJSON, err := json.Marshal(&retrieveVaultRequest{
+		Vsk: base64.RawStdEncoding.EncodeToString(vsk),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := http.Post(
+		v.vaultAPIURL+"/vaults/"+uuid,
+		"application/json",
+		bytes.NewBuffer(reqJSON),
+	)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	body, err := parseAPIResponse(resp)
+	if err != nil {
+		return nil, err
+	}
+
+	var parsedBody retrieveVaultResponse
+	err = json.Unmarshal(body, &parsedBody)
+	if err != nil {
+		return nil, err
+	}
+
+	// Decrypt encrypted vault file
+	encVfBase64 := parsedBody.EncryptedVault
+	encVf, err := base64.RawStdEncoding.DecodeString(encVfBase64)
+	if err != nil {
+		return nil, err
+	}
+
+	vf, err := decrypt(encVf, vk)
+	if err != nil {
+		return nil, err
+	}
+
+	var items []VaultItem
+	err = json.Unmarshal(vf, &items)
+	if err != nil {
+		return nil, err
+	}
+
+	return items, nil
+}
+
+func (v *vault) computeVk(uuid string, pass string, version VkVersion) []byte {
+	// In the future, we can increase iterations by doing a switch and mapping
+	// version to a given amount of iterations.
+
+	// version = V1 defaults to 100.000 iterations
+	iterations := 100000
+
+	return pbkdf2.Key([]byte(pass), []byte(string(version)+v.vaultSaltSecret+uuid), iterations, vaultKeyLength, sha512.New)
+}
+
+func (v *vault) computeVsk(vk []byte, pass string, version VkVersion) []byte {
+	iterations := 100000
+
+	return pbkdf2.Key(vk, []byte(string(version)+v.vaultSaltSecret+pass), iterations, vaultKeyLength, sha512.New)
+}
+
+func encrypt(data []byte, key []byte) ([]byte, error) {
+	block, err := aes.NewCipher(key[:])
+	if err != nil {
+		return nil, err
+	}
+
+	gcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return nil, err
+	}
+
+	nonce := make([]byte, gcm.NonceSize())
+	_, err = io.ReadFull(rand.Reader, nonce)
+	if err != nil {
+		return nil, err
+	}
+
+	return gcm.Seal(nonce, nonce, data, nil), nil
+}
+
+func decrypt(ciphertext []byte, key []byte) ([]byte, error) {
+	block, err := aes.NewCipher(key[:])
+	if err != nil {
+		return nil, err
+	}
+
+	gcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(ciphertext) < gcm.NonceSize() {
```

Comment:
@dmerrill6 

Only `<` or `!=` ?

Reply:
I think `<` is correct since, if I understand correctly, the nonce is prepended to the beginning of the ciphertext therefore it's fine if `len(ciphertext) >= nonce size`